File: Binary Tree - DFS/Medium/437. Path Sum III/path_sum_III.cpp
//Approach: Brute-Force
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int pathSum(TreeNode* root, int targetSum) {
        if (root == nullptr)
            return 0;
        return dfs(root, targetSum) +            
           pathSum(root->left, targetSum) +  
           pathSum(root->right, targetSum);
  }
    int dfs(TreeNode* root, long long targetSum) {
        if (root == nullptr)
            return 0;
            // cout << "Visiting node " << root->val << " with targetSum: " << targetSum << endl;
        return (targetSum == root->val) +                
           dfs(root->left, targetSum - root->val) +  
           dfs(root->right, targetSum - root->val);
    }
};



//Approach: Hash Map
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int pathSum(TreeNode* root, int targetSum) {
        unordered_map<long long, int> prefixSumCount;
        prefixSumCount[0] = 1;  // Base case: prefix sum = 0 has occurred once
        return dfs(root, 0LL, targetSum, prefixSumCount);
  }

  int dfs(TreeNode* node, long long currSum, int targetSum, unordered_map<long long, int>& prefixSumCount) {
    if (!node)
      return 0;

    currSum += node->val;
    int numPathsToCurr = 0;
        if (prefixSumCount.count(currSum - targetSum)) {
            numPathsToCurr = prefixSumCount[currSum - targetSum];
        }
    prefixSumCount[currSum]++;  // add current sum to map

    int leftPaths = dfs(node->left, currSum, targetSum, prefixSumCount);
    int rightPaths = dfs(node->right, currSum, targetSum, prefixSumCount);

    prefixSumCount[currSum]--;  // backtrack (remove this sum count for other paths)
    
    return numPathsToCurr + leftPaths + rightPaths;
    }
};

File: Binary Tree - DFS/Medium/437. Path Sum III/path_sum_III.py
#Approach: Brute-Force
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        if not root:
            return 0
        return (
            self.dfs(root, targetSum)
            + self.pathSum(root.left, targetSum)
            + self.pathSum(root.right, targetSum)
        )

    def dfs(self, node: TreeNode, targetSum: int) -> int:
        if not node:
            return 0
        return (
            (1 if node.val == targetSum else 0)
            + self.dfs(node.left, targetSum - node.val)
            + self.dfs(node.right, targetSum - node.val)
        )



#Approach: Hash Map
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        prefix_sum_count = defaultdict(int)
        prefix_sum_count[0] = 1  # Base case

        return self.dfs(root, 0, targetSum, prefix_sum_count)

    def dfs(self, node, curr_sum, target, prefix_sum_count):
        if not node:
            return 0

        curr_sum += node.val
        num_paths = prefix_sum_count[curr_sum - target]

        # Add current prefix sum to map
        prefix_sum_count[curr_sum] += 1

        # Recurse into left and right subtrees
        left = self.dfs(node.left, curr_sum, target, prefix_sum_count)
        right = self.dfs(node.right, curr_sum, target, prefix_sum_count)

        # Backtrack
        prefix_sum_count[curr_sum] -= 1

        return num_paths + left + right
